File: SPOJ/cpp/BYTESM2.c
#include<stdio.h>
#define MAX(X, Y) (((X) > (Y)) ? (X) : (Y))

int B[2][105];
int main(){
	int t;
	scanf("%d",&t);
	while(t--){
		int n,m,i,j,aij,ret;
		scanf("%d%d",&n,&m);
		for(j=0;j<=m+1;j++){
			B[0][j]=0;
			B[1][j]=0;
		}
		ret = 0;
		for(i=0;i<n;i++)
		{
			for(j=1;j<=m;j++)
			{
				scanf("%d",&aij);
				B[i%2][j] = aij+MAX(B[(i+1)%2][j-1],MAX(B[(i+1)%2][j],B[(i+1)%2][j+1]));
				ret = MAX(ret,B[i%2][j]);
			}

		}
		printf("%d\n",ret);	
	}
}

File: SPOJ/java/BYTESM2.java
import java.io.*;
import java.util.*;
import static java.lang.Math.max;
import static java.lang.Integer.parseInt;

class Main{

	public static void main(String[] args) throws IOException{	

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		PrintWriter out = new PrintWriter(System.out);
		int t = parseInt(br.readLine());
		while(t--!=0){
			StringTokenizer st = new StringTokenizer(br.readLine());
			int n = parseInt(st.nextToken());
			int m = parseInt(st.nextToken());
			int[][] B = new int[2][m+2];
			int ret = 0;
			for(int i=0;i<n;i++)
			{
				st = new StringTokenizer(br.readLine());
				for(int j=1;j<=m;j++)
				{
					int aij = parseInt(st.nextToken());
					B[i%2][j] = aij+max(B[(i+1)%2][j-1],max(B[(i+1)%2][j],B[(i+1)%2][j+1]));
					ret = max(ret,B[i%2][j]);
				}
			}
			out.printf("%d\n",ret);
		}
		out.flush();
		out.close();
		br.close();
	}

}

